components/places/src/storage/history.rs

```diff
@@ -53,6 +53,11 @@ pub fn apply_observation_direct(
     if url.as_str().len() > super::URL_LENGTH_MAX {
         return Ok(None);
     }
+    if let Some(ref preview_image_url) = visit_ob.preview_image_url {
```

Comment:
Like a few lines above, we want to parse the URL here to ensure it's valid and canonical, before checking the length. Also, for the URL we just drop the observation entirely, which you also do here - I'd think we'd be better off just dropping the preview image URL in this case (ie, still record the observation, just with a null preview image?)

I think we should fail on an invalid URL rather than dropping it, just like we do for the url itself - ie, "invalid URL" == error, "too long url" == "silent ignore"